src/pykeen/nn/emb.py

```diff
@@ -580,3 +581,246 @@ def forward(
+        """
+        Compose two batches of vectors.
+
+        The tensors have to be broadcastable.
+
+        :param a: shape: s_1
+            The first tensor.
+        :param b: shape: s_2
+            The second tensor.
+        :return: shape: s
+        """
+        raise NotImplementedError
+
+
+class SubtractionCompositionFunction(CompositionFunction):
+    def forward(self, a: torch.FloatTensor, b: torch.FloatTensor) -> torch.FloatTensor:
+        return a - b
+
+
+class MultiplicationCompositionFunction(CompositionFunction):
+    def forward(self, a: torch.FloatTensor, b: torch.FloatTensor) -> torch.FloatTensor:
+        return a * b
+
+
+class CrossCorrelationCompositionFunction(CompositionFunction):
+    def forward(self, a: torch.FloatTensor, b: torch.FloatTensor) -> torch.FloatTensor:
+        raise NotImplementedError
+
+
+composition_resolver = Resolver.from_subclasses(
+    CompositionFunction,
+    default=MultiplicationCompositionFunction,
+)
+
+
+class CompGCNLayer(nn.Module):
+    """A single layer of the CompGCN model."""
+
+    def __init__(
+        self,
+        input_dim: int,
+        output_dim: int,
+        dropout: float = 0.0,
+        use_bias: bool = True,
+        use_relation_bias: bool = False,
+        composition: Hint[CompositionFunction] = None,
+        activation: Hint[nn.Module] = nn.Identity,
+        activation_kwargs: Optional[Mapping[str, Any]] = None,
+    ):
+        super().__init__()
+
+        # entity-relation composition
+        self.composition = composition_resolver.make(composition)
+
+        # message passing weights
+        self.w_loop = nn.Parameter(data=torch.empty(input_dim, output_dim))
+        self.w_fwd = nn.Parameter(data=torch.empty(input_dim, output_dim))
+        self.w_bwd = nn.Parameter(data=torch.empty(input_dim, output_dim))
+
+        # linear relation transformation
+        self.w_rel = nn.Linear(in_features=input_dim, out_features=output_dim, bias=use_relation_bias)
+
+        # layer-specific self-loop relation representation
+        self.self_loop = nn.Parameter(data=torch.empty(1, input_dim))
+
+        # other components
+        self.drop = nn.Dropout(dropout)
+        self.bn = nn.BatchNorm1d(output_dim)
+        self.bias = Bias(output_dim) if use_bias else None
+        self.activation = activation_resolver.make(query=activation, pos_kwargs=activation_kwargs)
+
+        # initialize
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        """Reset the model's parameters."""
+        for w in (
+            self.w_loop,
+            self.w_fwd,
+            self.w_bwd,
+            self.w_rel,
+            self.self_loop,
+        ):
+            nn.init.xavier_uniform_(w)
+        self.bias.reset_parameters()
+
+    def message(
+        self,
+        x_e: torch.FloatTensor,
+        x_r: torch.FloatTensor,
+        triples: Optional[Tuple[torch.LongTensor, torch.LongTensor, torch.LongTensor]],
+    ) -> torch.FloatTensor:
```

Comment:
docstring please! what are x_e and x_r? How close is this to what PyTorch Geometric looks like?

Reply:
> How close is this to what PyTorch Geometric looks like?

ptg uses `edge_index` and `edge_type` of shape `(2, num_edges)` / `(num_edges,)`.

Reply:
I think there's a huge benefit to using the same interface as them, even if it means we need to do a few extra operations inside